### ISSUE TITLE
机器学习

### ISSUE THREAD

**Issue by hyoban** [hyoban/hyoban.github.io#69]
## 动态的 learning rate

1. loss 不再减小的时候，gradient 的大小并没有很小，很难走到 critical point。
2. 大的 learning rate 会一直震荡，小的 learning rate 在坡度很平滑的地方无法前进。
3. Root Mean Square，计算出来的 gradient，它的平方的平均再开根号。
4. RMS Prop，引入参数，可以手动指定当前的 gradient 的重要性。和前者相比，可以控制对新的 gradient 的灵敏程度，快速的调整 learning rate。
5. Adam就是RMS Prop加上Momentum。
6. 某个方向累积久了较小的 gradient 的话，就会引起 step 的变大，但还是能调整回来。所以需要加上 learning rate scheduling，有 decay 和 warmup 两种。
7. decay 让 learning rate 慢慢减小，使得后面 step 不会乱喷，然后慢慢走到终点。
8. warmup 起始会比较小，然后增大再减小。可能的解释是刚开始学到的 error surface 的情况没那么准确。
9. momentum 直接累加 gradient，而 Root Mean Square 只考虑 gradient 的大小，结果不会被抵消。

https://unclestrong.github.io/DeepLearning_LHY21_Notes/Notes_html/06_Adaptive%20Learning%20Rate.html

**Comment by hyoban** [comment]
## 参考课件 笔记

https://github.com/Fafa-DL/Lhy_Machine_Learning
https://github.com/unclestrong/DeepLearning_LHY21_Notes

**Comment by hyoban** [comment]
## Batch

1. 对数据进行 shuffle, 然后在分 batch, 使得每个 epoch 的 batch 都不一样.
2. 较小的 batch 的 epoch 会花费较长的时间, 对于每个 batch 来说, 由于并行处理, 时间差别不大.
3. 小的 batch 更新的方向比较 noisy, 但是 optimization 的效果会比较好.

**Comment by hyoban** [comment]
## Momentum

依据 gradient 的反方向和前一步移动的方向想加的结果来调整参数. 这样可以有助于我们离开 critical point.

**Comment by hyoban** [comment]
## 4 critical point

1. loss 不再下降的时候, 参数对于 loss 的微分为 0, 没有办法更新参数, 此时卡在 local minima 或者 saddle point. 统称为 critical point.
2. 我们有办法通过计算来判断具体的情况, 在 saddle point 的情况下, 我们依然可以更新参数.
3. 对于 error surface 的 local minima 来说, 在更高维的空间可能只是 saddle point 而已. 
4. 当训练中 loss 不再下降的时候, 实际情况中, 更常见的是 saddle point.

**Comment by hyoban** [comment]
## 3 General Guidance

![](https://gitee.com/unclestrong/deep-learning21_note/raw/master/imgbed/image-20210311205104719.png)

**Comment by hyoban** [comment]
## CNN

### intro

1. 专门用于影像
2. 输入时大小相同的照片, 输出是 one-hot 的向量对应种类, dimension 的长度决定识别出的种类个数
3. 图片是三位的 tensor, 一维代表图片的宽, 另一维代表高, 还有一维代表 channel 的数目 (RGB 三个channel)
4. 将三维的 tensor 拉直为一个向量, 每一维存的数值就是某个 pixel 某个颜色的强度, 由 RGB 三个颜色组成

### problem

过多的 weight 增加了 overfiting 的风险, 但是不是每个 neuron 对每个 input 的 dimension 都需要有 weight. 

### improvement 1

> 图像识别只需要看到某些 pattern 的话就可以分辨种类, pattern 通常都只占图像的一小部分, neuron 不需要看整个图片.

定义 Receptive field, 每个 neuron 只需要关系 rf 里面的事情. rf 可以依据自己的理解和需求来定义.

通常模式是: 

1. 看全部的 channel
2. 定义它的宽高为 kernel size. 
3. 对于每个 rf 有多个 neuron 来观察. 
4. 通过移动 stride 步来创建新的 rf. 
5. rf 是可以有重叠的空间. 
6. 对于超出范围, 设置 padding 补值. 

这样就能覆盖整张图片.

### improvement 2

> 相同的 pattern 出现在不同的位置, 不同的 neuron 做的事情就一样了, 只是 rf 不一样

parameter sharing, 使不同的 neuron 的 weight 完全一致, 但是 rf 不一样 (也就是输入不一样). 共享的参数被称为 filter

这样, Receptive Field 加上 Parameter Sharing, 就是 Convolutional Layer. 加上这些限制之后 cnn 的 model bias 就会比较大, 但这些限制是专门为了影像识别设置的, 会适合影像识别的任务, 对其他的任务就不一定了.

**Comment by hyoban** [comment]
## 7 classification

1. Classification 不能当作 Regression 来看, 让输出的种类为指定的数字. 因为这使得机器认为不同的种类之间存在着某种相关性
2. 处理分类的问题时, 常见的做法是将种类用 one-hot vector 来表示. 这样用距离来算它们之间就全是一样的大小
3. 处理输出时, 可以输出一个数值, 就可以使用相同的方法来输出一组数
4. 同时, 我们需要对输出加上一个 softmax 函数, 再去计算和准确值之间的距离, 将 y 中的任何值改为 0-1 之间
5. softmax 的另一个作用是使得大的和小的值之间的差距更大
6. 计算输出值和真实值之间的 loss 不只有一种方法, 但是不同的方法会影响我们训练模型的难易度
7. 常用的做法是使用 Cross-entropy, Make Minimize Cross-entropy 其实就是 maximize likelihood
8. 在pytorch 裡面, Cross-entropy 跟 Soft-max 是被绑在一起的
9. 选择 MAE 作为 loss function 的话, 很有可能 train 不起来. 因为在 loss 很大的时候, gradient 就已经很平整了, 没有办法用 gradient descent 走到 loss 小的地方
10. 总之, 选择好的 loss function 可以减少 train 的难度

[ref](https://unclestrong.github.io/DeepLearning_LHY21_Notes/Notes_html/08_Classification.html)

**Comment by hyoban** [comment]
## 8 Feature Normalization

1. 每一个 dimension 的值, 它的 scale 差距很大的时候, 就可能产生不同方向, 斜率非常不同, 坡度非常不同的 error surface
2. 让不同的 dimension 的值的范围相同, 就可以产生比较好训练的 error surface, 这就叫 Feature Normalization
3. 做法就是将每个 dimension 减去它们的 mean 然后除以 standard deviation, 让数值都变到 0 附近, variance 为 1
4. normalization 是每一层都要做, 至于放在 active function 之前还是之后都没太大影响. 如果是 sigmoid 的话, 比较推荐放在前面
5. 之后, network 不是处理单一的 example, 而是处理一堆 example. 每个 example 的改变都影响了其他的 example (影响了计算出来的 mean 和 standard deviation)
6. 实际上我们会考虑一个 batch, 所以这个方法又叫 batch normalization. 它适用于 batch size 比较大的时候, 这样才有的计算. 
7. 可以加上 y 和 b, 让 normalization 的 output 的 mean 不会一直被限制在 0, 给予比较好的自由度. 不过一般是在训练一段时间之后才加上的, 此时已经有比较好的 error surface, 所以 dimension 的分布有差别也没什么问题
8. testing 的时候用的 mean 和 standard deviation 就是在做 batch normalization 时候得到的 moving average. 不必等到组成一个 batch 的时候才开始计算
9. 因为 error surface 比较好训练, 所以我们可以把 learning rate 设置的大一些, 这样训练的效果会更好

**Comment by hyoban** [comment]
## RL

1. 不知道输出的形式, 适用于给数据做 label 很困难的时候
2. 在学习的过程中会和环境做互动, 得到 reward, 会知道当前的输出是好的还是不好的
3. Actor 和 environment 交互, 对 environment 给出的 observation, actor 做出 action, environment 给出 reward, 更新 observation, 然后一直重复

### RL 的步骤

1. RL 和普通的 ML 实际上都是三个步骤, 只是具体的定义方式不太相同
2. 定义 function 的输入输出时候, 输出定义为可能采取的 action 和对应的概率, 依照概率来随机采取 action, 而非直接采取最高概率的的 action
3. 定义 “loss” 为完成一整个 episode 之后得到的 total reward (return), 这就是我们最大化的目标
4. optimization: trajectory 定义为 actor 和 env 交互的序列, 每次得到 s 和 a, 通过 reward 函数得到 r, 需要让所有的 r 加起来最大. 这里 reward 的输出是和 action 以及 observation 都有关的.

### RL optimization 的问题

1. action 具有随机性
2. env 和 reward 都不是 network, env 是个黑盒子, 有可能有随机性, reward 就是个固定的规则而已
3. 和 GAN 不同的是, GAN 里面都是完全了解的 network

### Policy Gradient

1. 我们有办法控制 actor 去做或者不做我们指定的事情, 很类似在做 classification 的问题
2. 同时我们也可以控制在某个 observation 的情况下做或者不做的事情的程度, 想做还是一般想做

搜集上面需要的资料以及评价每个 action

1. 每个 action 会影响到接下来进行的行为
2. action 的 reward 有时候会在后面展示出来

真正 RL 的做法

1. 把每个 action 作用之后的所有 reward 加到一起
2. 后面很远的 reward 实际上和刚开始的 action 实际上关联没那么大, 需要加上 discount factor
3. reward 是相对的, 需要做标准化. 首先都要减去 baseline, 让 reward 有正有负

搜寻资料的问题

1. RL 训练的资料是自己每轮搜集, 然后更新参数之后需要重新搜集, 这样会比较花时间
2. 同样的 action 在不同的轮次对于不同的 actor 来说表现是不同的, 所以不能用之前的资料来训练
3. off-policy 可以有办法利用到之前搜集的资料, 分开 train 的 actor 和 interact 的 actor, 要知道两者之间的区别

actor 初始的随机性

1. 如果没有尝试过一些 action, 就没有办法知道 reward, 没办法 train 比较好的模型
2. 增加初始时的随机性, 可以尝试到多的 action
